### ISSUE TITLE
Eve Crash - Self test returning a non POWER_UP_SELF_TEST_PASSED status in wine

### ISSUE THREAD

**Issue by Caffe1neAdd1ct** [Algebro7/cryptopp#1]
First place i've started looking is inside fips140.h 

There are notes all through the file as to when and why the self test would pass or fail, would be interesting to narrow the reason down when it's running in Wine.

**Comment by Caffe1neAdd1ct** [comment]
These particular notes stand out to me:

```
/// \brief Performs the power-up self test
/// \param moduleFilename the fully qualified name of the module
/// \param expectedModuleMac the expected MAC of the components protected by the integrity check
/// \details Performs the power-up self test, and sets the self test status to
///   POWER_UP_SELF_TEST_PASSED or POWER_UP_SELF_TEST_FAILED.
/// \details The self tests for an algorithm are performed by the Algortihm class
///   when CRYPTOPP_ENABLE_COMPLIANCE_WITH_FIPS_140_2 is defined.
CRYPTOPP_DLL void CRYPTOPP_API DoPowerUpSelfTest(const char *moduleFilename, const byte *expectedModuleMac);
```

and 

```
/// \brief Performs the power-up self test on the DLL
/// \details Performs the power-up self test using the filename of this DLL and the
///   embedded module MAC, and sets the self test status to POWER_UP_SELF_TEST_PASSED or
///   POWER_UP_SELF_TEST_FAILED.
/// \details The self tests for an algorithm are performed by the Algortihm class
///   when CRYPTOPP_ENABLE_COMPLIANCE_WITH_FIPS_140_2 is defined.
CRYPTOPP_DLL void CRYPTOPP_API DoDllPowerUpSelfTest();
```

**Comment by Caffe1neAdd1ct** [comment]
So it sounds like this is an abstract defining an implementation which will be called depending on the crypto algo being called in the launcher at a guess..

**Comment by Caffe1neAdd1ct** [comment]
From your comments over on wine bugs:

Issue #1:

the game crashes at https://github.com/weidai11/cryptopp/blob/2d8e4680855691657cb8029ae5cb05c67c6dc654/cryptlib.cpp#L53 after checking the result value of a FIPS self-test. From what I can tell, DoPowerUpSelfTest() in fipstest.cpp is never called to run the tests, so this fails and it throws an exception. Just as a test case, I patched the cryptopp.dll that shipped with the launcher to remove that check and it gets past it just fine.

Failing on the third IF statement:

```
if (checkSelfTestStatus && FIPS_140_2_ComplianceEnabled())
	{
		if (GetPowerUpSelfTestStatus() == POWER_UP_SELF_TEST_NOT_DONE && !PowerUpSelfTestInProgressOnThisThread())
			throw SelfTestFailure("Cryptographic algorithms are disabled before the power-up self tests are performed.");

		if (GetPowerUpSelfTestStatus() == POWER_UP_SELF_TEST_FAILED)
			throw SelfTestFailure("Cryptographic algorithms are disabled after a power-up self test failed.");
	}
```

So that means checkSelfTestStatus and FIPS_140_2_ComplianceEnabled() are both true at this point and PowerUpSelfTestInProgressOnThisThread() must be coming back true as well, i'd suggest it's at least attempted to call the test?

**Comment by Caffe1neAdd1ct** [comment]
Hopefully some useful information:

https://github.com/Algebro7/cryptopp/blob/bc7d1bafa1e8ac732396374f0bca94ab9f396f1c/fipsalgt.cpp#L1210

https://bugs.winehq.org/show_bug.cgi?id=49543

https://gitlab.com/gnutls/gnutls/-/issues/894

I might be marching up the completely wrong path here but i'm guessing Wine doesn't support an algo expected by the library, failing the "self test". However I did see a note about proton wine already having a patch to fix this, which i've currently installed on this laptop to see if it was any better than wine-staging when first hitting this bug. 

Might be worth me testing once Wine have patched the msvcrt issue downstream on wine-valve / proton (https://aur.archlinux.org/packages/wine-valve/)

**Comment by Caffe1neAdd1ct** [comment]
Proton has support for it somehow. I've been looking into that and it seems possible to patch wine to support DH as well

It seems to use some internal functions mentioned in the response to the bug report you linked: https://lists.gnupg.org/pipermail/gnutls-help/2019-November/004580.html

The commit in Proton's github is this one https://github.com/ValveSoftware/wine/commit/8228cc1a058955bfaf0ba00f6259be76346a4651#diff-a60191c72c2892e5d0302344505e48e96ec53df295ad624cc4f0d37dd46b35e0

**Comment by Caffe1neAdd1ct** [comment]
Also might be worth mentioning this bug: https://bugs.winehq.org/show_bug.cgi?id=47843

Race conditions using bcrypt in other launchers / servers (Astroneer)

**Comment by Algebro7** [comment]
Thanks a lot for digging into this! There was another client update today and it...looks like it's not crashing anymore on the powerup self test? Stack trace from the latest (unpatched) build:

```
Backtrace:
=>0 0x000000007b010f56 EntryPoint+0xffffffffffffffff() in kernelbase (0x000000000084dab8)
  1 0x00000000429ba116 EntryPoint+0x820() in cryptopp (0x000000000084daf4)
  2 0x0000000042922338 EntryPoint+0xffffffffffffffff() in cryptopp (0x000000007b60dd08)
  3 0x0000000025ff5dec (0x000000008b55ff8b)
```

`0x42922338` in the cryptopp DLL is `SetNewAndDeleteFunctionPointers()` from dll.cpp, meaning we're now only crashing on the second spot. That's good news for us--not sure if they patched this on purpose or it was just a happy coincidence.

**Comment by Algebro7** [comment]
Actually scratch that, that address is the same I had in the original stack trace I posted on the forums. No idea what is going on with the FIPS stuff now...

I did see some comments indicating that setting breakpoints could cause the FIPS validation to fail, which if that's what's causing the powerupselftest to fail we won't have to do anything.

**Comment by Caffe1neAdd1ct** [comment]
` that setting breakpoints could cause the FIPS validation to fail`

is that something CCP would be doing in their launcher code and something we can't fix or?

**Comment by Algebro7** [comment]
I don't think so, everything I've seen so far is happening in the cryptopp dll, and I am hesitant to discuss what the Launcher binary is doing so I don't violate TOS. I was referring to this:

> At this point, we have a working DLL and a working DLL test program. Here's what running the DLL test program looks like. The failures are desired, and the FIPS self tests exits normally. If you set a software breakpoint in the image, then the tests will fail because a 0xCC is written which causes the CPU to trap. The new 0xCC instruction makes the integrity check fail.

From https://www.cryptopp.com/wiki/MSBuild_(Command_Line)

In any event, I can't get it to crash due to the failed powerupselftest anymore using the unpatched cryptopp dll that ships with the latest version of the launcher. So that problem may have magically taken care of itself *shrug*. [A PR](https://github.com/weidai11/cryptopp/pull/1062) is open to take care of the msvcrt problem on the cryptopp side but the maintainer said he's going to remove the DLL soon and that will hit windows users too, so the best course of action is to get CCP to talk to him so they can move to the statically linked version or the wrapper DLL.

**Comment by Caffe1neAdd1ct** [comment]
That makes sense :+1: 

So is it worth me building the latest version of Wine Staging with the patch at the bottom of https://bugs.winehq.org/show_bug.cgi?id=51465 to see if the msvcrt issue can be solved wine side for now? Hoping this would then work ok with the latest Eve Launcher as the `powerupselftest` issue has `magically taken care of itself shrug`  (gotta love those issues... )

**Comment by Algebro7** [comment]
> So is it worth me building the latest version of Wine Staging with the patch at the bottom of https://bugs.winehq.org/show_bug.cgi?id=51465 to see if the msvcrt issue can be solved wine side for now? 

Go for it! I was planning to give that a try too but this part seemed like it could take some time to figure out and I've been pretty busy:

> Then it'd be a matter of trying to compile it, fixing any instances where we rely on newer CRT features, and then doing the same thing with the tests.

**Comment by Caffe1neAdd1ct** [comment]
I think that is where i'll fall down, not a C++ / wine developer by any stretch of the imagination, i've managed to git bisect and compile to locate issues introduced at wine's end but i'm not fluent in any of this :/ 

I'll sit the laptop on the side and compile this tonight and see if any other errors crop up...

**Comment by Algebro7** [comment]
No problem, let me know how it goes and I'll work on it as well once I have some time. I've submitted a support ticket to CCP with the notes from the cryptopp maintainer as CCP Cemetary mentioned support tickets would be a possible way to get in touch with him. I'll let you know if I hear anything back.
